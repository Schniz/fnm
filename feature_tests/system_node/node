#!/bin/bash

echo "custom build"
