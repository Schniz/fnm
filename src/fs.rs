use std::path::Path;

#[cfg(unix)]
pub fn symlink_dir<P: AsRef<Path>, U: AsRef<Path>>(from: P, to: U) -> std::io::Result<()> {
    std::os::unix::fs::symlink(from, to)?;
    Ok(())
}

#[cfg(windows)]
pub fn symlink_dir<P: AsRef<Path>, U: AsRef<Path>>(from: P, to: U) -> std::io::Result<()> {
    junction::create(from, to)?;
    Ok(())
}

#[cfg(windows)]
pub fn remove_symlink_dir<P: AsRef<Path>>(path: P) -> std::io::Result<()> {
    std::fs::remove_dir(path)?;
    Ok(())
}

#[cfg(unix)]
pub fn remove_symlink_dir<P: AsRef<Path>>(path: P) -> std::io::Result<()> {
    std::fs::remove_file(path)?;
    Ok(())
}

pub fn shallow_read_symlink<P: AsRef<Path>>(path: P) -> std::io::Result<std::path::PathBuf> {
    std::fs::read_link(path)
}
