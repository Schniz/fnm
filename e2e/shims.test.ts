import { readFile } from "node:fs/promises"
import { join } from "node:path"
import { script } from "./shellcode/script.js"
import { Bash, Fish, PowerShell, WinCmd, Zsh } from "./shellcode/shells.js"
import testCwd from "./shellcode/test-cwd.js"
import describe from "./describe.js"
import testNodeVersion from "./shellcode/test-node-version.js"

for (const shell of [Bash, Zsh, Fish, PowerShell, WinCmd]) {
  describe(shell, () => {
    test(`outputs json`, async () => {
      const filename = `file.json`
      await script(shell)
        .then(
          shell.redirectOutput(
            shell.call("fnm", ["env", "--json", "--with-shims"]),
            {
              output: filename,
            }
          )
        )
        .takeSnapshot(shell)
        .execute(shell)

      if (shell.currentlySupported()) {
        const file = await readFile(join(testCwd(), filename), "utf8")
        expect(JSON.parse(file)).toEqual({
          FNM_ARCH: expect.any(String),
          FNM_DIR: expect.any(String),
          FNM_LOGLEVEL: "info",
          FNM_MULTISHELL_PATH: expect.any(String),
          FNM_NODE_DIST_MIRROR: expect.any(String),
          FNM_VERSION_FILE_STRATEGY: "local",
          FNM_VERSION_SWITCH_STRATEGY: "shims",
        })
      }
    })

    test(`runs Node through a shim`, async () => {
      await script(shell)
        .then(shell.env({ args: ["--with-shims"] }))
        .then(shell.call("fnm", ["install", "12.0.0"]))
        .then(shell.call("fnm", ["use", "12.0.0"]))
        .then(testNodeVersion(shell, "v12.0.0"))
        .takeSnapshot(shell)
        .execute(shell)
    })
  })
}
