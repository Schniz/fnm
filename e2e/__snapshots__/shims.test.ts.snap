// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bash outputs json: Bash 1`] = `
"set -e
fnm env --json --with-shims > file.json"
`;

exports[`Bash runs Node through a shim: Bash 1`] = `
"set -e
eval "$(fnm env --with-shims)"
fnm install 12.0.0
fnm use 12.0.0
if [ "$(node --version)" != "v12.0.0" ]; then
  echo "Expected node version to be v12.0.0. Got $(node --version)"
  exit 1
fi"
`;

exports[`Fish outputs json: Fish 1`] = `"fnm env --json --with-shims > file.json"`;

exports[`Fish runs Node through a shim: Fish 1`] = `
"fnm env --with-shims | source
fnm install 12.0.0
fnm use 12.0.0
set ____test____ (node --version)
if test "$____test____" != "v12.0.0"
  echo "Expected node version to be v12.0.0. Got $____test____"
  exit 1
end"
`;

exports[`PowerShell outputs json: PowerShell 1`] = `
"$ErrorActionPreference = "Stop"
fnm env --json --with-shims | Out-File file.json -Encoding UTF8"
`;

exports[`PowerShell runs Node through a shim: PowerShell 1`] = `
"$ErrorActionPreference = "Stop"
fnm env --with-shims | Out-String | Invoke-Expression
fnm install 12.0.0
fnm use 12.0.0
if ( "$(node --version)" -ne "v12.0.0" ) { echo "Expected node version to be v12.0.0. Got $(node --version)"; exit 1 }"
`;

exports[`Zsh outputs json: Zsh 1`] = `
"set -e
fnm env --json --with-shims > file.json"
`;

exports[`Zsh runs Node through a shim: Zsh 1`] = `
"set -e
eval "$(fnm env --with-shims)"
fnm install 12.0.0
fnm use 12.0.0
if [ "$(node --version)" != "v12.0.0" ]; then
  echo "Expected node version to be v12.0.0. Got $(node --version)"
  exit 1
fi"
`;
